vllm-ext/vllm/extension/ns/model/ns_model.py

```diff
@@ -0,0 +1,324 @@
+from vllm.sequence import SamplerOutput, SequenceGroupMetadata
+
+from vllm.logger import init_logger
+
+from vllm.extension import ns
+
+from vllm.extension.ns.model import _ModelPerf
+
+
+# import threading
+
+logger = init_logger(__name__)
+
+_NS_NUM_THREADS = "NS_NUM_THREADS"  # threads for prompt decoding
+_NS_NUM_THREADS_NEXT = "NS_NUM_THREADS_NEXT"  # threads for next token generation
+_NS_MAX_PROMPT_TOKENS = "NS_MAX_PROMPT_TOKENS"  # to be supported
+
+
+class NSModel(nn.Module):
+    def __init__(
+        self,
+        config: PretrainedConfig,
+        cache_config: Optional[CacheConfig] = None,
+        quant_config: Optional[QuantizationConfig] = None,
+        lora_config: Optional[LoRAConfig] = None,
+    ):
+        super(NSModel, self).__init__()
+        self.config = config
+        self.quant_config = quant_config
+        self.lora_config = lora_config
+        self.ie_model = None
+        self.tokenizer = None
+
+        self.padding_idx = config.pad_token_id
+        lora_vocab = (
+            (lora_config.lora_extra_vocab_size * (lora_config.max_loras or 1)) if lora_config else 0
+        )
+        self.vocab_size = config.vocab_size + lora_vocab
+        self.org_vocab_size = config.vocab_size
+        self.embed_tokens = VocabParallelEmbedding(
+            self.vocab_size,
+            config.hidden_size,
+            org_num_embeddings=config.vocab_size,
+        )
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        positions: torch.Tensor,
+        kv_caches: List[torch.Tensor],
+        attn_metadata: AttentionMetadata,
+    ):
+        assert len(kv_caches) == 1, "kv_caches should have 1 element here"
+        # use data_ptr and torch type in str to avoid inference engine model depends on pytorch and vllm types
+        # kv_cache pointer is set in NSBlockManager in advance
+        return self.ie_model(
+            input_ids.data_ptr(),
+            positions.data_ptr(),
+            attn_metadata.is_prompt,
+            attn_metadata.block_tables.data_ptr(),
+            attn_metadata.slot_mapping.data_ptr(),
+            attn_metadata.seq_lens,
+        )
+
+    def init_inference_engine(
+        self,
+        model_config: ModelConfig,
+        parallel_config: ParallelConfig,
+        scheduler_config: SchedulerConfig,
+    ):
+        # get available cores
+        try:
+            max_prompt_tokens = int(os.environ.get(_NS_MAX_PROMPT_TOKENS, "8192"))
+            # cpus_per_work is set to 1 for better ns perf so it's inappropriate to use ray to get available cores
+            # threads = ray.runtime_context.get_runtime_context().get_assigned_resources()['CPU']
+            physical_cores = psutil.cpu_count(logical=False)
+            # reserve one core for non-ns tasks
+            physical_cores = physical_cores if physical_cores <= 1 else physical_cores - 1
+            threads = int(os.environ.get(_NS_NUM_THREADS, str(physical_cores)))
```

Comment:
consider set the cores for ns according to `cpus_per_worker` in the inference config, all other predictors use this config to assign cpu resources.

Reply:
It's NS threads which is different from ray cpus. I want to separate them.

Reply:
Just came to my mind that bigger --num-cpus value causes ray starts up more background processes, like client.poll or server.poll processes which hurts overall performance. That's the reason why I set num-cpus to 1 to reduce them. Need to find an elegant way to balance them.